Remove option to configure custom config file via CONF_FILE or -Des.default.conf

brwe opened the pull request:
It is rarely used and was not consistently handled by different distributions anyway.
This commit also adds a test for specifying CONF_DIR when installing plugins and
starting elasticsearch.

relates to #12712  and #12954
closes #5329
closes #13715

Would be great if someone with a windows machine could test the bat file, I unfortunately can't.

brwe committed:
remove option to configure custom config file via CONF_FILE or -Des.default.conf

It is rarely used and was not consistently handled by different distributions anyway.
This commit also adds a test for specifying CONF_DIR when installing plugins and
starting elasticsearch.

relates to #12712 and #12954
closes #5329
closes #13715

rjernst commented on core/src/main/java/org/elasticsearch/node/internal/InternalSettingsPreparer.java:
```diff
@@ -85,24 +85,6 @@ public static Environment prepareEnvironment(Settings input, Terminal terminal)
-                output.loadFromPath(environment.configFile().resolve(System.getProperty("elasticsearch.config")));
-            }
-        }
         if (loadFromEnv) {
```
I think the local `loadFromEnv` and be removed?

rjernst commented on distribution/src/main/resources/bin/service.bat:
```diff
@@ -207,4 +207,8 @@ set /a conv=%conv% * 1024
 goto:eof
 
+:conffileset
+echo CONF_FILE setting is no longer supported. elasticsearch.yml must be placed in config folder and cannot be renamed.
```
in config -> in the config

rjernst commented on distribution/src/main/resources/bin/plugin:
```diff
@@ -59,6 +53,12 @@ if [ -n "$ES_ENV_FILE" ]; then
 
+# CONF_FILE setting was removed
+if [ ! -z "$CONF_FILE" ]; then
+    echo "CONF_FILE setting is no longer supported. elasticsearch.yml must be placed in config folder and cannot be renamed."
```
in config folder -> in the config directory

Note I suggest directory here instead of folder, since this is a nix script

rjernst commented on distribution/src/main/resources/bin/elasticsearch-systemd-pre-exec:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
```
in config folder -> in the config directory

rjernst commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -42,10 +42,25 @@
 # user has forgotten to run Maven to create a package.
+
+if [ `echo $@ | grep -c "es.default.conf" ` -gt 0 ]; then
+  echo "es.default.conf setting is no longer supported. elasticsearch.yml must be placed in config folder and cannot be renamed."
```
in config folder -> in the config directory

rjernst replied:
Ditto other places in this file.

rjernst commented on distribution/rpm/src/main/packaging/init.d/elasticsearch:
```diff
@@ -49,6 +49,12 @@ if [ -f "$ES_ENV_FILE" ]; then
 
+# CONF_FILE setting was removed
+if [ ! -z "$CONF_FILE" ]; then
+    echo "CONF_FILE setting is no longer supported. elasticsearch.yml must be placed in config folder and cannot be renamed."
```
in config folder -> in the config directory

rjernst commented on distribution/deb/src/main/packaging/init.d/elasticsearch:
```diff
@@ -93,10 +90,16 @@ if [ -f "$DEFAULT" ]; then
 fi
 
+# CONF_FILE setting was removed
+if [ ! -z "$CONF_FILE" ]; then
```
in config folder -> in the config directory

rjernst commented:
LGTM, I left a couple very minor comments.

nik9000 commented on qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash:
```diff
@@ -83,6 +83,56 @@ else
+@test "[$GROUP] start elasticsearch with a custom CONFIG_FILE and check failure" {
+    local CONF_FILE="$ESCONFIG/elasticsearch.yml"
+
+    if is_dpkg ; then
```
No space here I think.

nik9000 commented on qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash:
```diff
@@ -83,6 +83,56 @@ else
+        echo "CONF_FILE=$CONF_FILE" >> /etc/sysconfig/elasticsearch;
+    fi
+
+    if [ -f "/tmp/elasticsearch/bin/elasticsearch" ]; then
```
This part feels similar to start_elasticsearch_service in packaging_test_utils.bash - can you share any more code here?

nik9000 commented:
I also left some pretty minor comments.

electrical commented:
Where does the systemd pre-exec file come from ?

brwe commented:
@nik9000 @rjernst thanks for the review! addressed all comments.
@electrical I added the file to check if CONF_FILE was defined and not start elasticsearch if so. I don't know systemd too well so if there is another way to have this check before we start elasticsearch I'd be happy to change that.

brwe commented:
Chatted with @spinscale who wanted the check for `-Des` params to be moved from bin/elasticsearch script to Bootstrap. I added a commit for that too. I [reordered logging conf initialization](https://github.com/elastic/elasticsearch/pull/13772/files#diff-40fe9ad3b9a0769a26fe81051263be3dR240) so that we get a proper message, not sure though if this is problematic?

brwe commented:
@rjernst or @nik9000 or @spinscale  can you take another look?

brwe commented:
Forgot to add that I tried out the service.bat on windows now too. Thanks @dliappis for the help!

nik9000 commented:
LGTM

nik9000 commented on qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash:
```diff
@@ -239,18 +239,62 @@ clean_before_test() {
+# $1 expected status code
+# $2 additional command line args
+run_elasticsearch_service() {
+    # Set the CONF_DIR setting in case we start as a service
```
I'd probably save $1 and $2 to named variables because they can get confusing lower down.

nik9000 commented on qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash:
```diff
@@ -239,18 +239,62 @@ clean_before_test() {
 [ -f /usr/share/java/jayatanaag.jar ] && export JAVA_TOOL_OPTIONS="-javaagent:/usr/share/java/jayatanaag.jar"
 # And now we can start Elasticsearch normally, in the background (-d) and with a pidfile (-p).
-/tmp/elasticsearch/bin/elasticsearch -d -p /tmp/elasticsearch/elasticsearch.pid
+timeout 60s /tmp/elasticsearch/bin/elasticsearch $BACKGROUND -p /tmp/elasticsearch/elasticsearch.pid -Des.path.conf="$CONF_DIR" $2
```
The timeout feels funky when we are backgrounding elasticsearch. Maybe don't use it there?

nik9000 commented on qa/vagrant/src/test/resources/packaging/scripts/packaging_test_utils.bash:
```diff
@@ -262,31 +306,11 @@ BASH
-    elif is_sysvinit; then
-        run service elasticsearch status
-        [ "$status" -eq 0 ]
+        [ "$status" -eq $1 ]
```
This would be clearer is `$1` were `$expected_status` or something.

brwe committed:
name variables and don't always use the timeout

brwe commented:
@nik9000 addressed all comments. I made the timeout now dependent on if we run in background or not but my bash is that of a 3 year old so if you have any suggestion to improve readability don't hold back.

nik9000 commented:
LGTM

bstascavage commented:
This was actually a feature we used and this required me to re-architecture our cluster deployment.  I have no idea why, in 2015, you would remove the ability to specify a config file, and instead for the user to have a config file be a specific filename.  Obviously the damage is done, but I would dispute the claim that it is 'It is rarely used'
